pkg/util/cache/intern.go

```diff
@@ -0,0 +1,359 @@
+	minFileSize    int64
+	maxPerInterner int
+	lock           sync.Mutex
+}
+
+// NewKeyedStringInterner creates a Keyed String Interner with a max per-origin quota of maxQuota
+func NewKeyedStringInterner(cfg cconfig.Component) *KeyedInterner {
+	stringInternerCacheSize := cfg.GetInt("dogstatsd_string_interner_size")
+	enableMMap := cfg.GetBool("dogstatsd_string_interner_mmap_enable")
+
+	if enableMMap {
+		closeOnRelease := !cfg.GetBool("dogstatsd_string_interner_mmap_preserve")
+		tempPath := cfg.GetString("dogstatsd_string_interner_tmpdir")
+		minSizeKb := cfg.GetInt("dogstatsd_string_interner_mmap_minsizekb")
+		maxStringsPerInterner := cfg.GetInt("dogstatsd_string_interner_origin_max_strings")
+		return NewKeyedStringInternerVals(stringInternerCacheSize, closeOnRelease, tempPath, minSizeKb, maxStringsPerInterner)
+	}
+	return NewKeyedStringInternerMemOnly(stringInternerCacheSize)
+}
+
+// NewKeyedStringInternerVals takes args explicitly for initialization
+func NewKeyedStringInternerVals(stringInternerCacheSize int, closeOnRelease bool, tempPath string, minFileKb, maxStringsPerInterner int) *KeyedInterner {
+	cache, err := lru.NewWithEvict(stringInternerCacheSize, func(_, internerUntyped interface{}) {
+		interner := internerUntyped.(*stringInterner)
+		interner.Release(1)
+	})
+	if err != nil {
+		return nil
+	}
+	return &KeyedInterner{
+		interners:      cache,
+		maxQuota:       -1,
+		lastReport:     time.Now(),
+		tmpPath:        tempPath,
+		minFileSize:    int64(minFileKb * 1024),
+		maxPerInterner: maxStringsPerInterner,
+		closeOnRelease: closeOnRelease,
+	}
+}
+
+// NewKeyedStringInternerMemOnly is a memory-only cache with no disk needs.
+func NewKeyedStringInternerMemOnly(stringInternerCacheSize int) *KeyedInterner {
+	cache, err := lru.NewWithEvict(stringInternerCacheSize, func(_, internerUntyped interface{}) {
+		interner := internerUntyped.(*stringInterner)
+		interner.Release(1)
+	})
+	if err != nil {
+		return nil
+	}
+	return &KeyedInterner{
+		interners:      cache,
+		maxQuota:       -1,
+		lastReport:     time.Now(),
+		tmpPath:        "",
+		maxPerInterner: initialInternerSize,
+		closeOnRelease: false,
+	}
+}
+
+// NewKeyedStringInternerForTest is a memory-only cache with a small default size.  Useful for
+// most tests.
+func NewKeyedStringInternerForTest() *KeyedInterner {
+	return NewKeyedStringInternerMemOnly(512)
+}
+
+var sGlobalQueryCount = atomic.NewInt64(0)
+var sFailedInternalCount = atomic.NewInt64(0)
+
+// LoadOrStoreString interns a string for an origin
+func (i *KeyedInterner) LoadOrStoreString(s string, origin string, retainer InternRetainer) string {
+	if Check(s) {
+		return i.LoadOrStore(unsafe.Slice(unsafe.StringData(s), len(s)), origin, retainer)
+	}
+	sFailedInternalCount.Add(1)
+	return "<invalid>"
+}
+
+// LoadOrStore interns a byte-array to a string, for an origin
+func (i *KeyedInterner) LoadOrStore(key []byte, origin string, retainer InternRetainer) string {
+	sGlobalQueryCount.Add(1)
```

Comment:
It's not clear to me that this needs to be atomic. We increment here only -- and you've got it atomic to skip taking the lock that's implicit in `loadOrStore` -- but the sum is only used in a debug log. I guess, how accurate does this need to be? This is going to be a sequentially consistent operation: we're forcing all CPUs to sync, then we skip some takes of the mutex, then we force all the CPUs to sync again.

Reply:
I'm using a guideline of atomic operations on MESIF 'E' states being ~8ns in the common (l1 uncontested) case  https://arxiv.org/pdf/2010.09852.pdf

I think a good part of that 8ns is still the regular unexclusive op.

Generally, I just wanted rough statistics without too much bother :)

Reply:
Fair, but in a large piece of software like the Agent as core counts ramp it's not always clear that atomics cheap in the micro are cheap in the macro. A nit flag since we don't have data one way or the other yet and there's other low handing sync issues in the Agent yet.